File: NB/nb_layers.pl
# ===== Output inventory info for layer 1 =====
# LAYER 1 has LAYER RANK 1 and attributes defined by the various L1 fields L1S1 L1PR1, L1HT, etc.
# LAYER 2 has LAYER RANK 2 and attributes defined by the various L2 fields L2S1 L2PR1, L2HT, etc.
$NUMBER_OF_LAYERS = 1;
$StandStructureCode = "S";
if ($Forested != 0) { 
	# Only these should be forested stands
	if ($Canopy >=1) {  #L1VS >= 1
		if ($UCanopy ==0 || isempty($UCanopy)) {  
			#  (L2VS ==0 !! is.null(L2VS)) {
 			$NUMBER_OF_LAYERS = 1;
			$StandStructureCode = "S";
		} else { 
			#L2Vs > 0
 			$NUMBER_OF_LAYERS = 2;
 			if ($Canopy >1 && $UCanopy >1) { 
				#(L1VS > 1 && L2VS > 1)
   				$StandStructureCode = "C";  
			} else {
  				$StandStructureCode = "M"; 
			}
		}
	} elsif($UCanopy == 0 || isempty($UCanopy)) { # L1VS == 0 or NULL && L2VS == 0 or NULL
		if (!isempty($Sp1)) { #(L1S1 != NULL)
			if (!isempty($USp1)) { #(L2S1 != NULL) # We have two layers anyway.
   				$NUMBER_OF_LAYERS =2;
   				$StandStructureCode = "M";
 			} else {
   				$NUMBER_OF_LAYERS =  1;
   				$StandStructureCode = "S";
 			}
		} else {
			$NUMBER_OF_LAYERS =1;
			$particular=1;
			$StandStructureCode = "S";
			#LAYER 1 has LAYER RANK 1, but origin, height, density and species composition are "Missing"
			#If there is no .DST record, create one with type and other attributes coded Missing
		}
	} else {
		$keys="l1VS==0 and L2VS !=0"."#".$Canopy."#".$UCanopy;
		$herror{$keys}++;
	}
}

File: NB/nb_layers.R
num_of_layers = 1
stand_structure = "S"
if (fst>0) {
	if (l1vs>0) {
		if (l2vs==0 | is.na(l2vs) {
 			num_of_layers = 1
		    stand_structure = "S"
		} else { # l2vs > 0
 			num_of_layers = 2
 			if (l1vs>1 & l2vs>1) { 
   				stand_structure = "C"  
			} else {
  				stand_structure = "M"
			}
		}
	} else if (l1vs==0 & (l2vs==0 | is.na(l2vs)) {
		if (!is.na(l1s1)) {
			if (!is.na(l2s1)) {
   				num_of_layers = 2
   				stand_structure = "M"
 			} else {
   				num_of_layers = 1
   				stand_structure = "S"
 			}
		} else { # LAYER 1 has LAYER RANK 1, but origin, height, density and species composition are "Missing"
			num_of_layers = 1
			particular = 1
			stand_structure = "S"
			
		}
	} else {
        # for example: l1vs==0 & l2vs>0 ?
        print("error code...")
	}
} else {

}

Revised extraction of rules from Perl code (fixing a problem detected in Perl code):

Cases where L1VS > 0 and L2VS = 0:
if l1vs > 0 & (l2vs==0 | is.na(l2vs)) { stand_structure="S" & num_of_layers=1 }
if l1vs > 1 & l2vs > 1 { stand_structure="C" & num_of_layers=2 }
if l1vs > 0 & l2vs > 0 { stand_structure="M" & num_of_layers=2 }

Cases where L1VS = 0 and L2VS > 0:
perl code missing for these cases...

Cases where both L1VS and L2VS are 0 but there are species:
if (l1vs==0 | is.na(l1vs1)) & (l2vs==0 | is.na(l2vs)) & !is.na(l1s1) & !is.na(l2s1) { stand_structure="M" & num_of_layers=2 }
if (l1vs==0 | is.na(l1vs1)) & (l2vs==0 | is.na(l2vs)) & !is.na(l1s1) & is.na(l2s1) { stand_structure="S" & num_of_layers=1 }
if (l1vs==0 | is.na(l1vs1)) & (l2vs==0 | is.na(l2vs)) & !is.na(l1s1) { stand_structure="S" & num_of_layers=1 }
...the last line seems redundant